pkg/analyze/data_source.go

```diff
@@ -0,0 +1,116 @@
+package analyze
+
+import (
+	"context"
+	"fmt"
+	scyllaversioned "github.com/scylladb/scylla-operator/pkg/client/scylla/clientset/versioned"
+	scyllav1listers "github.com/scylladb/scylla-operator/pkg/client/scylla/listers/scylla/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/pager"
+)
+
+type DataSource struct {
+	PodLister            corev1listers.PodLister
+	ServiceLister        corev1listers.ServiceLister
+	SecretLister         corev1listers.SecretLister
+	ConfigMapLister      corev1listers.ConfigMapLister
+	ServiceAccountLister corev1listers.ServiceAccountLister
+	ScyllaClusterLister  scyllav1listers.ScyllaClusterLister
+}
+
+func BuildListerWithOptions[T any](
+	ctx context.Context,
+	factory func(cache.Indexer) T,
+	listFunc func(ctx context.Context, options metav1.ListOptions) (runtime.Object, error),
+	options metav1.ListOptions,
+) (T, error) {
+	p := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+		return listFunc(ctx, opts)
+	}))
+
+	// Prevent users from providing unwanted ones or tempering options that pager controls
+	options = metav1.ListOptions{
+		LabelSelector: options.LabelSelector,
+		FieldSelector: options.FieldSelector,
+	}
+
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{"NamespaceIndex": cache.MetaNamespaceIndexFunc})
+	err := p.EachListItemWithAlloc(ctx, options, func(obj runtime.Object) error {
+		err := indexer.Add(obj)
+		if err != nil {
+			return fmt.Errorf("can't add object to indexer %v: %w", obj, err)
+		}
+		return nil
+	})
+	if err != nil {
+		return *new(T), fmt.Errorf("can't iterate over list items: %w", err)
+	}
+
+	return factory(indexer), nil
+}
+
+func BuildLister[T any](ctx context.Context, factory func(cache.Indexer) T, listFunc func(ctx context.Context, options metav1.ListOptions) (runtime.Object, error)) (T, error) {
+	return BuildListerWithOptions[T](ctx, factory, listFunc, metav1.ListOptions{})
+}
+
+func NewDataSourceFromClients(
+	ctx context.Context,
+	kubeClient kubernetes.Interface,
+	scyllaClient scyllaversioned.Interface,
+) (DataSource, error) {
```

Comment:
nit: you wouldn't have to return an empty struct on an error if the return type was (*DataSource, error), which also seems more idiomatic for the status return pattern

Reply:
done